### PR TITLE
Switch to using http.Server.Shutdown

### PR DESCRIPTION
**Pull request by jameshartig**

### CODE REVIEW

File: genapi/listener_reloader.go [hunk 1]
```diff
@@ -58,9 +61,21 @@ func (lr *listenerReloader) Accept() (net.Conn, error) {
 
 // Close only closes the wrapping net.Listener, it's assumed the underlying one
 // will be subsequently closed down the chain
+// In the event of http.Server.Shutdown, Close will be called by the defer in
+// Serve and the Shutdown function. So we detect if we're already closed and
+// return an error if we already are. This seems consistant with builtin.
 func (lr *listenerReloader) Close() error {
-	close(lr.closeCh)
-	return (<-lr.lch).Close()
+	// If we are already closed then there won't be anything to read on lch,
```
Comment (mediocregopher) [inline genapi/listener_reloader.go]: There's still _kind of_ a race condition with this I think:

* Process A calls Close(), reads a listener off lch
* Process B calls Accept(), reads a listener off lch, calls Accept() on the listener
* Process A (still in Close) calls Close() on that listener
* something weird happens, cause I dunno if you can call Accept() and Close() on a listener at the same time.

What we could do is have closeCh be a `chan net.Listener` which spin is always trying to write to, and once it's written to it spin assumes it should return out.

File: genapi/genapi.go [hunk 2]
```diff
@@ -887,6 +889,7 @@ func (g *GenAPI) doLever() {
 		g.Lever.Add(lever.Param{
 			Name:        "--unhealthy-timeout",
 			Description: "Number of milliseconds to appear unhealthy after a stop signal is received",
+			Default:     "1",
```
Comment (mediocregopher) [inline genapi/genapi.go:898]: What's this for?

Comment (jameshartig) [reply genapi/genapi.go:898]: I removed the random waiting for skyapi and just defaulted this.

File: genapi/listener_reloader.go (hunk 1 above)
Comment (jameshartig) [reply genapi/listener_reloader.go]: I'll try writing to closeCh. Theoretically that should work

File: genapi/listener_reloader.go [hunk 3]
```diff
@@ -58,9 +68,16 @@ func (lr *listenerReloader) Accept() (net.Conn, error) {
 
 // Close only closes the wrapping net.Listener, it's assumed the underlying one
 // will be subsequently closed down the chain
+// In the event of http.Server.Shutdown, Close will be called by the defer in
+// Serve and the Shutdown function. So we detect if we're already closed and
+// return an error if we already are. This seems consistant with builtin.
 func (lr *listenerReloader) Close() error {
-	close(lr.closeCh)
-	return (<-lr.lch).Close()
+	select {
```
Comment (mediocregopher) [inline genapi/listener_reloader.go]: sooooooooo still a problem here. If spin doesn't happen to be in the select at the moment Close is called the default case here might hit instead. If instead we did close the closeCh, when you read from it here you can do:
```
l, ok := <-lr.closeCh
if !ok {
    return errors.New("already closed")
}
return l.Close()
```
and skip the select entirely

**Pull request merged**